src/MarginTrading.Backend.Services/Services/CfdCalculatorService.cs

```diff
@@ -59,13 +61,64 @@ public decimal GetQuoteRateForQuoteAsset(string accountAssetId, string assetPair
             return 1.0M / quote.Ask;
         }
 
+        public decimal GetFplRate(string accountAssetId, string instrumentId, string legalEntity, bool fplSign)
+        {
+            var assetPair = _assetPairsCache.GetAssetPairById(instrumentId);
+            
+            if (accountAssetId == assetPair.QuoteAssetId)
+                return 1;
+
+            var assetPairQuoteAccount = _assetPairsCache.TryFindAssetPairStraight(assetPair.QuoteAssetId, accountAssetId, legalEntity);
+            IAssetPair assetPairAccountQuote = null;
+            if (assetPairQuoteAccount == null)
+                assetPairAccountQuote = _assetPairsCache.TryFindAssetPairStraight(accountAssetId, assetPair.QuoteAssetId, legalEntity);
+
+            var rate = fplSign
+                ? assetPairQuoteAccount != null
+                    ? _quoteCacheService.GetQuote(assetPairQuoteAccount.Id).Ask
+                    : 1 / _quoteCacheService.GetQuote(assetPairAccountQuote.Id).Bid
+                : assetPairQuoteAccount != null
+                    ? _quoteCacheService.GetQuote(assetPairQuoteAccount.Id).Bid
+                    : 1 / _quoteCacheService.GetQuote(assetPairAccountQuote.Id).Ask;
+            
+            return rate;
+        }
+
+        public decimal GetSwapRate(string accountAssetId, string instrumentId, string legalEntity, bool swapSign)
+        {
+            var assetPair = _assetPairsCache.GetAssetPairById(instrumentId);
+            
+            if (accountAssetId == assetPair.BaseAssetId)
+                return 1;
+
+            var assetPairBaseAccount = _assetPairsCache.TryFindAssetPairStraight(assetPair.BaseAssetId, accountAssetId, legalEntity);
+            IAssetPair assetPairAccountBase = null;
+            if (assetPairBaseAccount == null)
+                assetPairAccountBase = _assetPairsCache.TryFindAssetPairStraight(accountAssetId, assetPair.BaseAssetId, legalEntity);
+
+            var rate = swapSign
+                ? assetPairBaseAccount != null
+                    ? _quoteCacheService.GetQuote(assetPairBaseAccount.Id).Ask
+                    : 1 / _quoteCacheService.GetQuote(assetPairAccountBase.Id).Bid
+                : assetPairBaseAccount != null
+                    ? _quoteCacheService.GetQuote(assetPairBaseAccount.Id).Bid
+                    : 1 / _quoteCacheService.GetQuote(assetPairAccountBase.Id).Ask;
+            
+            return rate;
+        }
+
+        public decimal GetMarginRate(string accountAssetId, string instrumentId, string legalEntity)
+        {
+            return GetSwapRate(accountAssetId, instrumentId, legalEntity, true);
```

Comment:
??? )
Here is the best illustration of problems with such methods for every margin parameter.
We already have GetQuoteRateForBaseAsset and GetQuoteRateForQuoteAsset.
And they are almost not used now.
Why you did not changed them and used where you need rate for quote or for base asset?

Reply:
Changed GetQuoteRateForBaseAsset and GetQuoteRateForQuoteAsset, removed 3 separate methods.